server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -368,7 +362,25 @@ protected int getRefCount(ElasticsearchDirectoryReader reference) {
         }
 
         @Override
-        protected void decRef(ElasticsearchDirectoryReader reference) throws IOException { reference.decRef(); }
+        protected void decRef(ElasticsearchDirectoryReader reference) throws IOException {
+            reference.decRef();
+        }
+    }
+
+    @Override
+    final boolean assertSearcherIsWarmedUp(String source, SearcherScope scope) {
+        if (scope == SearcherScope.EXTERNAL) {
+            switch (source) {
+                // these sources can access the external searchers before the engine is warmed up
+                case "segments":
+                case "segments_stats":
+                case "refresh_needed":
```

Comment:
why is this the case? This is only called by `scheduledRefresh`, which requires the shard to be started or post_recovery. We explicitly refresh however before moving to post_recovery state (see postRecovery method).

Reply:
Good catch :). Fixed in https://github.com/elastic/elasticsearch/pull/48605/commits/204faaae77f565cc7d3521e4dfa61b60fdb4f75f